lib/fuse_lowlevel.c

```diff
@@ -2890,16 +2890,38 @@ int fuse_session_mount(struct fuse_session *se, const char *mountpoint)
 			close(fd);
 	} while (fd >= 0 && fd <= 2);
 
-	/* Open channel */
-	fd = fuse_kern_mount(mountpoint, se->mo);
-	if (fd == -1)
-		return -1;
-	se->fd = fd;
+	if (!*mountpoint) {
+		/*
+		 * Pre-mounted mode: To allow FUSE helpers to run without
+		 * privileges, the caller may open /dev/fuse and mount the file
+		 * system before launching the helper. The helper doesn't need
+		 * to handle mounting in this case, which is reflected by the
+		 * mountpoint parameter being the empty string. The /dev/fuse FD
+		 * is expected to be open on file descriptor 0 in that case.
+		 */
+		struct stat st;
+		if (fstat(STDIN_FILENO, &st) == -1 || !S_ISCHR(st.st_mode) ||
+		    isatty(STDIN_FILENO)) {
+			fprintf(stderr ,"fuse: File descriptor 0 must be open on /dev/fuse for pre-mounted mode.\n");
+			return -1;
+		}
+		se->fd = STDIN_FILENO;
```

Comment:
Note to self: Using STDIN might not be the best idea, since it'll cause problems with FUSE binaries that expect to read input from the user (sshfs for example).